src/java.base/share/classes/jdk/crac/impl/CriticalUnorderedContext.java

```diff
@@ -0,0 +1,90 @@
+package jdk.crac.impl;
+
+import jdk.crac.CheckpointException;
+import jdk.crac.Context;
+import jdk.crac.Resource;
+import jdk.crac.RestoreException;
+
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+/**
+ * Context performing Checkpoint notification in unspecified order.
+ * Concurrent registration along beforeCheckpoint notification triggers
+ * immediate notification on being registered resource.
+ * @param <R>
+ */
+public class CriticalUnorderedContext<R extends Resource> extends AbstractContext<R> {
+    private final WeakHashMap<R, Void> resources = new WeakHashMap<>();
+    private ExceptionHolder<CheckpointException> concurrentCheckpointException = null;
+
+    private synchronized List<R> snapshot() {
+        return this.resources.entrySet().stream()
+            .map(Map.Entry::getKey)
+            .toList();
+    }
+
+    @Override
+    protected List<R> checkpointSnapshot() {
+        return snapshot();
+    }
+
+    @Override
+    protected List<R> restoreSnapshot() {
+        // return updated set, as registration has called beforeCheckpoint()
+        return snapshot();
+    }
+
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws CheckpointException {
+        synchronized (this) {
+            concurrentCheckpointException = new ExceptionHolder<>(CheckpointException::new);
+        }
+
+        try {
+            super.beforeCheckpoint(context);
+        } catch (CheckpointException e) {
+            synchronized (this) {
+                concurrentCheckpointException.handle(e);
+            }
+        }
+        synchronized (this) {
+            ExceptionHolder<CheckpointException> e = concurrentCheckpointException;
+            concurrentCheckpointException = new ExceptionHolder<>(CheckpointException::new);
+            e.throwIfAny();
+        }
+    }
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws RestoreException {
+        CheckpointException racedException;
+        synchronized (this) {
+            racedException = concurrentCheckpointException.getIfAny();
+            concurrentCheckpointException = null;
+        }
+
+        ExceptionHolder<RestoreException> restoreException = new ExceptionHolder<>(RestoreException::new);
+        restoreException.handle(racedException);
+
+        restoreException.runWithHandler(() -> {
+            super.afterRestore(context);
+        });
+
+        restoreException.throwIfAny();
+    }
+
+    @Override
+    public void register(R resource) {
+        synchronized (this) {
+            resources.put(resource, null);
+            if (concurrentCheckpointException != null) {
+                try {
+                    invokeBeforeCheckpoint(resource);
+                } catch (Exception e) {
+                    concurrentCheckpointException.handle(e);
```

Comment:
This makes sense. As a thought excercise, it is possible to have add a last Resource, that will query all pending exceptions in all previous contextes, or make this context block once all beforeCheckpoint are done and there is nowhere to report exceptions. Both are also possible for user contextes, outside of the system classes.

Reply:
If you're speaking about querying Contexts, that means extending its API for traversing and fetching those. Alternatively, should the contexts report exceptions to a well-known resource this resource would become a part of the API.

Reply:
It is possible to add query method to JDK contextes without changing public context interface. The same way, the alternative won't require changes in the public API (it will be a convention between JDK contextexes and the well-known resource).

Reply:
Sure, if we're talking only about JDKContext and this is used nowhere except in JDKContext the options are wider, and your proposed solution would work.